Refactor strings column factories

harrism opened the pull request:
This PR refactors strings column factories to eliminate the use of `device_vector` and `std::vector` parameters, and to facility more use of `device_uvector` in calls to the factories. This is a small part of #7287 . Multiple versions of `make_strings_columns` take `device_vector` parameters. This PR expands the use of iterator and `device_span` versions to enable switching to `device_uvector` as described in #7287.  It also adds new `make_device_uvector_async/sync` utility functions.

This will help facilitate safe CUDA stream usage.

davidwendt commented on cpp/include/cudf/detail/utilities/vector_factories.hpp:
```diff
@@ -0,0 +1,242 @@
+ * @param c The input device container from which to copy
+ * @param stream The stream on which to allocate memory and perform the copy
+ * @param mr The memory resource to use for allocating the returned device_uvector
+ * @return rmm::device_uvector<T> A device_uvector containing the copied data
```
```suggestion
 * @return  A device_uvector containing the copied data
```
Don't include the return type in `@return` doxygen tag.

harrism replied:
Ugh, my vscode plugin does that automatically. :(  Will remove.

davidwendt commented on cpp/include/cudf/detail/utilities/vector_factories.hpp:
```diff
@@ -0,0 +1,242 @@
+
+/**
+ * @brief Asynchronously construct a `device_uvector` containing a deep copy of data from a
+ * device_span
```
```suggestion
 * `device_span`
```

davidwendt commented on cpp/include/cudf/strings/detail/strings_column_factories.cuh:
```diff
@@ -25,6 +25,7 @@
 
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
+#include "rmm/thrust_rmm_allocator.h"
```
```suggestion
#include <rmm/thrust_rmm_allocator.h>
```

davidwendt commented on cpp/src/strings/strings_column_factories.cu:
```diff
@@ -18,92 +18,23 @@
 #include <rmm/device_vector.hpp>
 #include <rmm/exec_policy.hpp>
 
 #include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
 #include <thrust/transform_reduce.h>
```
```suggestion
```
These headers are not used by this `.cu` file.

davidwendt commented on cpp/src/strings/strings_column_factories.cu:
```diff
@@ -18,92 +18,23 @@
 #include <cudf/column/column_factories.hpp>
 #include <cudf/detail/nvtx/ranges.hpp>
 #include <cudf/detail/valid_if.cuh>
```
```suggestion
```
This header is not needed anymore.

davidwendt commented on cpp/src/strings/strings_column_factories.cu:
```diff
@@ -18,92 +18,23 @@
 #include <cudf/column/column_factories.hpp>
 #include <cudf/detail/nvtx/ranges.hpp>
```
You removed all of the `CUDF_FUNC_RANGE();` calls to the `detail::make_strings_column`. Usually these are put in the non-detail function right before calling the detail version.

harrism commented:
@davidwendt I addressed your feedback and also added some more device_vector -> uvector changes in strings (the simple ones I found). I think that's enough for this PR though.

davidwendt commented on cpp/src/strings/findall.cu:
```diff
@@ -164,29 +164,28 @@ std::unique_ptr<table> findall_re(
   for (int32_t column_index = 0; column_index < columns; ++column_index) {
-    rmm::device_vector<string_index_pair> indices(strings_count);
-    string_index_pair* d_indices = indices.data().get();
+    rmm::device_uvector<string_index_pair> indices(strings_count, stream);
```
I'm making some more involved changes in this file in #7428

harrism replied:
Resolved.

davidwendt commented on cpp/src/strings/split/split.cu:
```diff
@@ -506,8 +506,8 @@ std::unique_ptr<table> split_fn(strings_column_view const& strings_column,
-  rmm::device_vector<string_index_pair> tokens(columns_count * strings_count);
-  string_index_pair* d_tokens = tokens.data().get();
+  rmm::device_uvector<string_index_pair> tokens(columns_count * strings_count, stream);
+  string_index_pair* d_tokens = tokens.data();
```
I'm making more involved changes to this file in #7427

harrism replied:
Resolved (there weren't conflicts in this file).

davidwendt reviewed (changes requested):
Some of the changes in this PR are more than necessary for refactoring the strings column factory. Two of the non-refactory changes become a conflict for at least two of my PRs. Request maybe this PR can be broken up.

jrhemstad commented on cpp/include/cudf/column/column_factories.hpp:
```diff
@@ -344,20 +345,19 @@ std::unique_ptr<column> make_fixed_width_column(
  */
 std::unique_ptr<column> make_strings_column(
-  const rmm::device_vector<thrust::pair<const char*, size_type>>& strings,
+  cudf::device_span<thrust::pair<const char*, size_type>> strings,
```
I think you want a `const` on the pair type too. 
```suggestion
  cudf::device_span<thrust::pair<const char*, size_type> const> strings,
```

harrism replied:
Added, thanks.

jrhemstad commented on cpp/include/cudf/strings/detail/strings_column_factories.cuh:
```diff
@@ -95,6 +95,50 @@ std::unique_ptr<column> make_strings_column(IndexPairIterator begin,
+  size_type bytes         = std::distance(chars_begin, chars_end) * sizeof(char);
+  if (strings_count == 0) return strings::detail::make_empty_strings_column(stream, mr);
+
+  CUDF_EXPECTS(null_count < strings_count, "null strings column not yet supported");
```
Really?

jrhemstad commented on cpp/src/io/json/reader_impl.cu:
```diff
@@ -600,9 +601,22 @@ table_with_metadata reader::impl::convert_data_to_table(rmm::cuda_stream_view st
+                                    0,
+                                    stream);
+  auto repl   = make_strings_column(cudf::detail::make_device_uvector_async(repl_chars, stream),
+                                  cudf::detail::make_device_uvector_async(repl_offsets, stream),
```
Is there a sync somewhere later to make sure this copy is done?

harrism replied:
Ignoring the later D2H copy without a stream, I don't think this requires a sync, since the function we are inside takes a stream, and is thus assumed to follow stream ordering semantics on that stream (and does so by passing the stream to these uvector and make_strings_column functions.

jrhemstad replied:
But the `target_chars/target_offsets` and `repl_chars/repl_offsets` are temporary local vectors that will get destroyed when the function exits and the copy may not have completed yet. Or am I missing something?

harrism replied:
Changing the structure is out of scope of this PR, so I added a sync for now with a comment explaining why.

jrhemstad commented on cpp/src/io/json/reader_impl.cu:
```diff
@@ -600,9 +601,22 @@ table_with_metadata reader::impl::convert_data_to_table(rmm::cuda_stream_view st
+                                  0,
+                                  stream);
 
   thrust::host_vector<cudf::size_type> h_valid_counts = d_valid_counts;
```
This is doing a D2H copy w/o a stream.

harrism replied:
There are many other bugs that this PR does not fix as well! :)

harrism replied:
This function also uses a log of device_vectors. Replacing them is non-trivial (unlike some of the strings changes I made), so I prefer to tackle that in a separate PR.

jrhemstad commented on cpp/include/cudf/detail/utilities/vector_factories.hpp:
```diff
@@ -0,0 +1,242 @@
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  rmm::device_uvector<T> ret(source_data.size(), stream, mr);
+  CUDA_TRY(cudaMemcpyAsync(ret.data(),
+                           source_data.data(),
+                           source_data.size() * sizeof(T),
+                           cudaMemcpyDefault,
+                           stream.value()));
+  stream.synchronize();
+  return ret;
```
This can just call the `async` version and then call `sync`, right?

harrism replied:
Yeah good point.

harrism replied:
Done.

jrhemstad commented on cpp/include/cudf/detail/utilities/vector_factories.hpp:
```diff
@@ -0,0 +1,242 @@
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  rmm::device_uvector<T> ret(source_data.size(), stream, mr);
+  CUDA_TRY(cudaMemcpyAsync(ret.data(),
+                           source_data.data(),
+                           source_data.size() * sizeof(T),
+                           cudaMemcpyDefault,
+                           stream.value()));
+  stream.synchronize();
+  return ret;
```
Same here. Use the async version to implement the sync version.

harrism replied:
Done.

jrhemstad commented on cpp/include/cudf/detail/utilities/vector_factories.hpp:
```diff
@@ -0,0 +1,242 @@
+ * @param mr The memory resource to use for allocating the returned device_uvector
+ * @return A device_uvector containing the copied data
+ */
+template <template <typename, typename...> typename Container,
+          typename T,
+          std::enable_if_t<std::is_convertible<Container<T>, host_span<T const>>::value>* = nullptr>
+rmm::device_uvector<T> make_device_uvector_async(
+  Container<T> const& c,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  return make_device_uvector_async(host_span<T const>{c}, stream, mr);
+}
```
This is a slightly cleaner way to do it. 
```suggestion
template <typename Container,
          std::enable_if_t<std::is_convertible<Container, host_span<Container::value_type const>>::value>* = nullptr>
rmm::device_uvector<Container::value_type> make_device_uvector_async(
  Container const& c,
  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
{
  return make_device_uvector_async(host_span<Container::value_type const>{c}, stream, mr);
}
```

harrism replied:
Marginally, since you have to use the `typename` keyword everywhere you have `Container::value_type` it's just about the same verbosity.  But I made the change.

harrism commented:
> Some of the changes in this PR are more than necessary for refactoring the strings column factory. Two of the non-refactory changes become a conflict for at least two of my PRs. Request maybe this PR can be broken up.

The goal of the PR is ultimately to replave device_vector... If you just merge your PRs first I will resolve conflicts.

davidwendt commented on cpp/include/cudf/column/column_factories.hpp:
```diff
@@ -330,7 +331,7 @@ std::unique_ptr<column> make_fixed_width_column(
 
 /**
- * @brief Construct STRING type column given a device vector of pointer/size pairs.
+ * @brief Construct STRING type column given a device span of pointer/size pairs.
```
Add a blank line after the `@brief` tag.
```suggestion
 * @brief Construct STRING type column given a device span of pointer/size pairs.
 
```
https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/docs/DOCUMENTATION.md#brief

harrism replied:
I think the blank line needs to start with a `*`.

davidwendt replied:
Oops, right.

davidwendt commented on cpp/include/cudf/column/column_factories.hpp:
```diff
@@ -344,20 +345,19 @@ std::unique_ptr<column> make_fixed_width_column(
 
 /**
- * @brief Construct STRING type column given a device vector of string_view.
+ * @brief Construct STRING type column given a device span of string_view.
```
```suggestion
 * @brief Construct STRING type column given a device span of string_view.
 
```

davidwendt commented on cpp/include/cudf/column/column_factories.hpp:
```diff
@@ -372,118 +372,68 @@ std::unique_ptr<column> make_strings_column(
- * and null mask and null count. The columns and mask are moved into the
- * resulting strings column.
+ * @brief Constructs a STRING type column given offsets column, chars columns, and null mask and
+ *  null count. The columns and mask are moved into the resulting strings column.
```
```suggestion
 *  null count. 
 *
 * The columns and mask are moved into the resulting strings column.
```

davidwendt commented on cpp/tests/strings/combine_tests.cpp:
```diff
@@ -281,8 +281,13 @@ TEST_F(StringsConcatenateWithColSeparatorTest, SingleColumnStringMixNoReplacemen
     {"", "", "", "bbabc", "", "d", "éa", "", "bbb", "éééf"},
     {false, false, true, true, false, true, true, false, true, true});
 
+  CUDA_TRY(cudaDeviceSynchronize());
```
Why is this needed here?

harrism replied:
Good catch, must be leftover from debugging

davidwendt commented on cpp/tests/strings/combine_tests.cpp:
```diff
@@ -281,8 +281,13 @@ TEST_F(StringsConcatenateWithColSeparatorTest, SingleColumnStringMixNoReplacemen
   auto results =
     cudf::strings::concatenate(cudf::table_view{{col0}}, cudf::strings_column_view(sep_col));
+
+  CUDA_TRY(cudaDeviceSynchronize());
```
I don't think this is needed either.

harrism commented:
@cwharris what changes are you requesting?

davidwendt commented on cpp/tests/strings/combine_tests.cpp:
```diff
@@ -283,6 +283,7 @@ TEST_F(StringsConcatenateWithColSeparatorTest, SingleColumnStringMixNoReplacemen
 
   auto results =
     cudf::strings::concatenate(cudf::table_view{{col0}}, cudf::strings_column_view(sep_col));
+
```
Maybe revert this file. You are only adding a blank line now.

davidwendt commented on cpp/include/cudf/detail/utilities/vector_factories.hpp:
```diff
@@ -0,0 +1,237 @@
+/*
+ *
```
```suggestion

```

davidwendt reviewed (changes requested):
Some of the files need 2021 added to their copyright line.

harrism commented:
> Some of the files need 2021 added to their copyright line.

Thanks. Done.

shwina commented:
rerun tests

cwharris commented:
@harrism you addressed my concerns. :+1:

harrism commented:
@gpucibot merge
